### PR TITLE
WIP: Add Gaussian process classes

### PR DESCRIPTION
**Pull request by fonnesbeck**
This PR continues the work of @bwengals in implementing Gaussian processes. I've added a couple of mean functions here.  Classes for specifying Gaussian prior processes are forthcoming. As with the existing work, I've tried to follow the GPflow implementation pretty closely.

I have changed the covariance API a little here. I've replaced the `K` method with `__call__`, which seems to me to be cleaner, but have kept `K` around as an alias to avoid breakage.

This is not nearly ready, so please do not merge yet.

### CODE REVIEW

**Comment by bwengals** [comment]
Wow cool!  Good ```__call__```.  I was thinking there may be no need for mean fuction classes though, other than convenience.  A big selling point of PyMC3 for me is that any arbitrary theano code could be dropped in here, glms, keras, some theoretical model...

**Comment by fonnesbeck** [comment]
Yeah, that's a good point. I could imagine having the GPP default to the `Zero` mean function, which probably suits most users. 

Thinking of something like the following:

```
# Specify GP  covarance
K = Exponential(1, 0.1) + Linear(1, 0.5)

# Gaussian prior process
age_effect = GPP('age_effect`, K)

# Sampling error
Sigma = tt.eye(n)*10

# Multivariate normal likelihood at observed points where X=age_obs
likelihood = MvNormal('likelihood', age_effect(age_obs), Sigma, observed=y)
```

**Comment by bwengals** [comment]
I agree completely.  The usage you describe is exactly what I am thinking!  In my PR I got hung up on making ```sample_ppc``` work (being nonparametric X and Z both need to get in there somehow) and punted.  The mean would go  
```
m = ...
age_effect = GPP('age_effect', mean=m, cov=K)
```

File: pymc3/gp/cov.py [hunk 1]
```diff
@@ -148,10 +145,11 @@ def __init__(self, input_dim, lengthscales, alpha, active_dims=None):
         self.lengthscales = lengthscales
         self.alpha = alpha
 
-    def K(self, X, Z=None):
+    def __call__(self, X, Z=None):
         X, Z = self._slice(X, Z)
         return tt.power((1.0 + 0.5 * self.square_dist(X, Z) * (1.0 / self.alpha)), -1.0 * self.alpha)
 
+    K = __call__
```
Comment (twiecki) [inline pymc3/gp/cov.py]: Since this is so new, I don't think we need backwards compat.

Comment (bwengals) [reply pymc3/gp/cov.py]: I'm happy to change tutorials and docs to use ```__call__``` instead of ```.K(X)``` asap

Comment (twiecki) [reply pymc3/gp/cov.py]: Should definitely do that. You could do a PR onto this branch.

**Comment by fonnesbeck** [comment]
Strange CI failure due to distutils an numpy?

**Comment by fonnesbeck** [comment]
I've been through a few iterations locally, and what seems to work best is an API such as the following:

```
with pm.Model() as gp_fit:
    
    M = pm.gp.mean.Zero()
    
    ρ = pm.Gamma('ρ', 1, 1)
    σ = pm.HalfCauchy('σ', 2.5)
    
    K = pm.gp.cov.Matern32(1, ρ)
    
    gp = pm.gp.GP('gp', M, K, sigma=tt.eye(N)*σ, observed={'X':X, 'Y':y})
```

This keeps the X grid associated with the GP in the model. It is a little more black box than what I proposed above, however. I have not pushed anything yet, but please feel free to provide feedback.

**Comment by fonnesbeck** [comment]
This is just a prototype. I tried a couple different implementations before committing this one, but I'm not sure if its the best approach. Might be nice to separate the likelihood from the GPP, etc. Open to suggestions.

For example, we could just stick with a set of covariance and mean functions, with a `MvNormal` likelihood, which would require only a convenience function for drawing predictive values.

**Comment by bwengals** [comment]
Yeah.. a lot of ways this could go! I will keep thinking about it.  Right now I've been working on callback functions to print the progress of ```find_MAP``` in a nice way (another thing I really love about GPy).  It's an alright exercise for understanding pymc3s internals.  

I worry that computing predictions during sampling (like I did in the tutorial) is too slow.  And your original proposal for GP's signature was super clean and math-notation-like.  Maybe there could be two separate GP classes, one for the conjugate normal case kinda like your recent commit, and another that is more general and could be used in the nonconjugate case.  ~~In the latter maybe we can get away with having X be a shared...~~ not sure have to think about it / give it a try.

**Comment by fonnesbeck** [comment]
I think you are right regarding calculating predictions. This should really occur after sampling, as we do for posterior predictive checks.

**Comment by fonnesbeck** [comment]
Forget it. Rebase hell. Closing and starting again.

**Comment by twiecki** [comment]
You're really bad at this :P.

**Comment by fonnesbeck** [comment]
I guess so. I've been rebasing for over a decade, so I'm not sure why I'm running into these issues now.

**Pull request closed without merging**